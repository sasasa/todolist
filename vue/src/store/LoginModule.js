import { buildModuleTypes } from '@/helpers/store';
// import fetch from 'cross-fetch'
import 'cross-fetch/polyfill';

const URL = process.env.VUE_APP_API_URL_BASE

const moduleName = 'LoginModule';


export const namespaced = true;

export const state = {
  errors: [
  ],
  signup_errors: {
  },
  reset_password_errors: {
  },
  signup_success: false,
  send_email_success: false,
  reset_password_success: false,
  user: {
    authentication_token: '',
    password: '',
    email: '',
    id: 0,
  },
  login: false,
};

const GETTER_TYPES = {
  USER: 'USER',
  LOGIN: 'LOGIN',
  ERRORS: 'ERRORS',
  SIGNUP_ERRORS: 'SIGNUP_ERRORS',
  SIGNUP_SUCCESS: 'SIGNUP_SUCCESS',
  SEND_EMAIL_SUCCESS: 'SEND_EMAIL_SUCCESS',
  RESET_PASSWORD_SUCCESS: 'RESET_PASSWORD_SUCCESS',
  RESET_PASSWORD_ERRORS: 'RESET_PASSWORD_ERRORS',
};
const MUTATION_TYPES = {
  USER: 'USER',
  LOGIN: 'LOGIN',
  ERRORS: 'ERRORS',
  SIGNUP_ERRORS: 'SIGNUP_ERRORS',
  SIGNUP_SUCCESS: 'SIGNUP_SUCCESS',
  SEND_EMAIL_SUCCESS: 'SEND_EMAIL_SUCCESS',
  RESET_PASSWORD_SUCCESS: 'RESET_PASSWORD_SUCCESS',
  RESET_PASSWORD_ERRORS: 'RESET_PASSWORD_ERRORS',
};
const ACTION_TYPES = {
  LOGIN_ASYNC: 'LOGIN_ASYNC',
  SIGNUP_ASYNC: 'SIGNUP_ASYNC',
  LOGOUT_ASYNC: 'LOGOUT_ASYNC',
  SEND_EMAIL_ASYNC: 'SEND_EMAIL_ASYNC',
  RESET_PASSWORD_ASYNC: 'RESET_PASSWORD_ASYNC',
};

export const LOGIN_GETTER_TYPES = buildModuleTypes({
  moduleName,
  types: GETTER_TYPES,
});
export const LOGIN_MUTATION_TYPES = buildModuleTypes({
  moduleName,
  types: MUTATION_TYPES,
});
export const LOGIN_ACTION_TYPES = buildModuleTypes({
  moduleName,
  types: ACTION_TYPES,
});

export const getters = {
  [GETTER_TYPES.USER](state) {
    return state.user
  },
  [GETTER_TYPES.LOGIN](state) {
    return state.login
  },
  [GETTER_TYPES.ERRORS](state) {
    return state.errors;
  },
  [GETTER_TYPES.SIGNUP_ERRORS](state) {
    return state.signup_errors;
  },
  [GETTER_TYPES.SIGNUP_SUCCESS](state) {
    return state.signup_success;
  },
  [GETTER_TYPES.SEND_EMAIL_SUCCESS](state) {
    return state.send_email_success
  },
  [GETTER_TYPES.RESET_PASSWORD_SUCCESS](state) {
    return state.reset_password_success
  },
  [GETTER_TYPES.RESET_PASSWORD_ERRORS](state) {
    return state.reset_password_errors
  },
};

export const mutations = {
  [MUTATION_TYPES.USER](state, { user }) {
    state.user = user
  },
  [MUTATION_TYPES.LOGIN](state, { login }) {
    state.login = login
  },
  [MUTATION_TYPES.ERRORS](state, { errors }) {
    state.errors = errors
  },
  [MUTATION_TYPES.SIGNUP_ERRORS](state, { errors }) {
    state.signup_errors = errors
  },
  [MUTATION_TYPES.SIGNUP_SUCCESS](state, { success }) {
    state.signup_success = success
  },
  [MUTATION_TYPES.SEND_EMAIL_SUCCESS](state, { success }) {
    state.send_email_success = success
  },
  [MUTATION_TYPES.RESET_PASSWORD_SUCCESS](state, { success }) {
    state.reset_password_success = success
  },
  [MUTATION_TYPES.RESET_PASSWORD_ERRORS](state, { errors }) {
    state.reset_password_errors = errors
  },
};

export const actions = {
  [ACTION_TYPES.RESET_PASSWORD_ASYNC](context, payload) {
    context.commit('toggle_loading', payload, { root: true })
    context.commit(MUTATION_TYPES.RESET_PASSWORD_SUCCESS, { success: false })

    fetch(`${URL}/users/reset_password.json`, {
      method: 'POST',
      mode: 'cors',
      credentials: 'include',
      headers: { 
        'X-CSRF-Token': localStorage.csrf,
        'Content-type': 'application/json',
      },
      body: JSON.stringify(payload),
    })
    .then(res => res.json())
    .then(data => {
      if (data.success) {
        context.commit(MUTATION_TYPES.RESET_PASSWORD_SUCCESS, { success: true })
      } else {
        context.commit(MUTATION_TYPES.RESET_PASSWORD_SUCCESS, { success: false })
        context.commit(MUTATION_TYPES.RESET_PASSWORD_ERRORS, { errors: data.errors })
      }
      context.commit('toggle_loading', payload, { root: true })
    })
    .catch((error) => {
      // console.error(error)
      alert(error)
      context.commit(MUTATION_TYPES.RESET_PASSWORD_SUCCESS, { success: false })
      context.commit('toggle_loading', payload, { root: true })
    })
  },
  [ACTION_TYPES.SEND_EMAIL_ASYNC](context, payload) {
    context.commit('toggle_loading', payload, { root: true })
    context.commit(MUTATION_TYPES.SEND_EMAIL_SUCCESS, { success: false })

    fetch(`${URL}/users/send_email.json`, {
      method: 'POST',
      mode: 'cors',
      credentials: 'include',
      headers: { 
        'X-CSRF-Token': localStorage.csrf,
        'Content-type': 'application/json',
      },
      body: JSON.stringify(payload),
    })
    .then(res => res.json())
    .then(data => {
      if (data.success) {
        context.commit(MUTATION_TYPES.SEND_EMAIL_SUCCESS, { success: true })
      } else {
        context.commit(MUTATION_TYPES.SEND_EMAIL_SUCCESS, { success: false })
      }
      context.commit('toggle_loading', payload, { root: true })
    })
    .catch((error) => {
      // console.error(error)
      alert(error)
      context.commit(MUTATION_TYPES.SEND_EMAIL_SUCCESS, { success: false })
      context.commit('toggle_loading', payload, { root: true })
    })
  },
  [ACTION_TYPES.LOGOUT_ASYNC](context, payload) {
    context.commit('toggle_loading', payload, { root: true })

    fetch(`${URL}/users/logout.json`, {
      method: 'POST',
      mode: 'cors',
      credentials: 'include',
      headers: { 
        'X-CSRF-Token': localStorage.csrf,
        'Content-type': 'application/json',
      },
      body: JSON.stringify(payload),
    })
    .then(res => res.json())
    .then(data => {
      if (data.success) {
        context.commit(MUTATION_TYPES.USER, { user: {} })
        context.commit(MUTATION_TYPES.LOGIN, { login: false })
      }
      context.commit('toggle_loading', payload, { root: true })
    })
    .catch((error) => {
      // console.error(error);
      alert(error)
      context.commit('toggle_loading', payload, { root: true })
    })
  },


  [ACTION_TYPES.LOGIN_ASYNC](context, payload) {
    context.commit('toggle_loading', payload, { root: true })

    fetch(`${URL}/users/login.json`, {
      method: 'POST',
      mode: 'cors',
      credentials: 'include',
      headers: { 'Content-type': 'application/json' },
      body: JSON.stringify(payload),
    })
    .then(res => res.json())
    .then(data => {
      if (data.errors) {
        const errors = data.errors
        context.commit(MUTATION_TYPES.ERRORS, { errors })
      } else {
        // LocalStorage に CSRF トークンを保存する
        localStorage.csrf = data.csrf;
  
        const user = data.user
        context.commit(MUTATION_TYPES.ERRORS, { errors: [] })
        context.commit(MUTATION_TYPES.USER, { user })
        context.commit(MUTATION_TYPES.LOGIN, { login: true })
      }
      context.commit('toggle_loading', payload, { root: true })
    })
    .catch((error) => {
      // console.error(error);
      alert(error)
      context.commit('toggle_loading', payload, { root: true })
    })
  },

  [ACTION_TYPES.SIGNUP_ASYNC](context, payload) {
    context.commit('toggle_loading', payload, { root: true })

    fetch(`${URL}/users.json`, {
      method: 'POST',
      mode: 'cors',
      credentials: 'include',
      headers: { 'Content-type': 'application/json' },
      body: JSON.stringify(payload),
    })
    .then(res => res.json())
    .then(data => {
      // alert(JSON.stringify(data))
      if (data.errors) {
        context.commit(MUTATION_TYPES.SIGNUP_SUCCESS, { success: false })
        context.commit(MUTATION_TYPES.SIGNUP_ERRORS, { errors: data.errors })
      } else {
        context.commit(MUTATION_TYPES.SIGNUP_SUCCESS, { success: true })
      }
      context.commit('toggle_loading', payload, { root: true })
    })
    .catch((error) => {
      // console.error(error);
      alert(error)
      context.commit('toggle_loading', payload, { root: true })
    })
  },
};