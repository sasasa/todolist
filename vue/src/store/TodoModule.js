import { buildModuleTypes } from '@/helpers/store';
// import fetch from 'cross-fetch'
import 'cross-fetch/polyfill';

const URL = process.env.VUE_APP_API_URL_BASE

const moduleName = 'TodoModule';


export const namespaced = true;

export const state = {
  todo: {
    content: '',
    due_on: '',
    finished: false,
    id: 0,
  },
  todos: [
  ],
  add_errors: {
    content: [],
  },
  add_success: false,
  update_errors: {
    content: [],
  },
  update_success: false,
};

const GETTER_TYPES = {
  TODOS: 'TODOS',
  TODO: 'TODO',
  ADD_ERRORS: 'ADD_ERRORS',
  ADD_SUCCESS: 'ADD_SUCCESS',
  UPDATE_ERRORS: 'UPDATE_ERRORS',
  UPDATE_SUCCESS: 'UPDATE_SUCCESS',
};
const MUTATION_TYPES = {
  GET_TODO: 'GET_TODO',
  UPDATE_TODO: 'UPDATE_TODO',
  GET_TODOS: 'GET_TODOS',
  REMOVE_TODO: 'REMOVE_TODO',
  ADD_TODO: 'ADD_TODO',
  ADD_ERRORS: 'ADD_ERRORS',
  ADD_SUCCESS: 'ADD_SUCCESS',
  UPDATE_ERRORS: 'UPDATE_ERRORS',
  UPDATE_SUCCESS: 'UPDATE_SUCCESS',
};
const ACTION_TYPES = {
  GET_TODO_ASYNC: 'GET_TODO_ASYNC',
  UPDATE_TODO_ASYNC: 'UPDATE_TODO_ASYNC',
  GET_TODOS_ASYNC: 'GET_TODOS_ASYNC',
  REMOVE_TODO_ASYNC: 'REMOVE_TODO_ASYNC',
  ADD_TODO_ASYNC: 'ADD_TODO_ASYNC',
};

export const TODO_GETTER_TYPES = buildModuleTypes({
  moduleName,
  types: GETTER_TYPES,
});
export const TODO_MUTATION_TYPES = buildModuleTypes({
  moduleName,
  types: MUTATION_TYPES,
});
export const TODO_ACTION_TYPES = buildModuleTypes({
  moduleName,
  types: ACTION_TYPES,
});

export const getters = {
  [GETTER_TYPES.TODOS](state) {
    return state.todos;
  },
  [GETTER_TYPES.TODO](state) {
    return state.todo;
  },
  [GETTER_TYPES.ADD_ERRORS](state) {
    return state.add_errors;
  },
  [GETTER_TYPES.ADD_SUCCESS](state) {
    return state.add_success;
  },
  [GETTER_TYPES.UPDATE_ERRORS](state) {
    return state.update_errors;
  },
  [GETTER_TYPES.UPDATE_SUCCESS](state) {
    return state.update_success;
  },
};

export const mutations = {
  [MUTATION_TYPES.GET_TODO](state, { todo }) {
    state.todo = todo
  },
  [MUTATION_TYPES.GET_TODOS](state, { todos }) {
    state.todos = todos
  },
  [MUTATION_TYPES.REMOVE_TODO](state, { todo }) {
    state.todos = state.todos.filter((t)=>{
      return t.id !== todo.id
    })
  },
  [MUTATION_TYPES.UPDATE_TODO](state, { todo }) {
    state.todos.forEach((t)=>{
      if(t.id === todo.id) {
        t.finished = todo.finished
        t.content = todo.content
        t.due_on = todo.due_on
      }
    })
  },
  [MUTATION_TYPES.ADD_TODO](state, { todo }) {
    state.todos.push(todo)
  },
  [MUTATION_TYPES.ADD_ERRORS](state, { errors }) {
    state.add_errors = errors
  },
  [MUTATION_TYPES.ADD_SUCCESS](state, { success }) {
    state.add_success = success
  },
  [MUTATION_TYPES.UPDATE_ERRORS](state, { errors }) {
    state.update_errors = errors
  },
  [MUTATION_TYPES.UPDATE_SUCCESS](state, { success }) {
    state.update_success = success
  },
};

export const actions = {
  [ACTION_TYPES.GET_TODO_ASYNC](context, payload) {
    context.commit('toggle_loading', payload, { root: true })
    fetch(`${URL}/todos/${payload.id}.json?user_email=${payload.user.email}`, { 
      method: 'GET',
      headers: {
        Authorization: `Bearer ${payload.user.authentication_token}`,
      }
    })
    .then(response => {
      if (200 <= response.status < 300 && response.ok) {
        return response.json();
      } else {
        throw new Error(`status NG ${response.status}:${response.statusText}`);
      }
    })
    .then(todo => {
      context.commit(MUTATION_TYPES.GET_TODO, { todo })
      // console.log(`--------------${ todo.id }`)
      context.commit('toggle_loading', payload, { root: true })
    })
    .catch((error) => {
      // console.error(error);
      
      alert(error)
      context.commit('toggle_loading', payload, { root: true })
    })
  },

  [ACTION_TYPES.GET_TODOS_ASYNC](context, payload) {
    context.commit('toggle_loading', payload, { root: true })
    fetch(`${URL}/todos.json?user_email=${payload.user.email}`, { 
      method: 'GET',
      headers: {
        Authorization: `Bearer ${payload.user.authentication_token}`,
      }
    })
    .then(response => {
      if (200 <= response.status < 300 && response.ok) {
        return response.json();
      } else {
        throw new Error(`status NG ${response.status}:${response.statusText}`);
      }
    })
    .then(todos => {
      context.commit(MUTATION_TYPES.GET_TODOS, { todos })
      context.commit('toggle_loading', payload, { root: true })
    })
    .catch((error) => {
      // console.error(error);
      alert(error)
      context.commit('toggle_loading', payload, { root: true })
    })
  },

  [ACTION_TYPES.REMOVE_TODO_ASYNC](context, payload) {
    context.commit('toggle_loading', payload, { root: true })
    fetch(`${URL}/todos/${payload.todo.id}.json?user_email=${payload.user.email}`, {
      method: 'DELETE',
      mode: 'cors',
      credentials: 'include',
      headers: { 
        'Content-type': 'application/json',
        'X-CSRF-Token': localStorage.csrf,
        Authorization: `Bearer ${payload.user.authentication_token}`,
      },
    })
    .then(response => {
      if (200 <= response.status < 300 && response.ok) {
        return response.json();
      } else {
        throw new Error(`status NG ${response.status}:${response.statusText}`);
      }
    })
    .then((todo) => {
      todo
      context.commit(MUTATION_TYPES.REMOVE_TODO, payload)
      context.commit('toggle_loading', payload, { root: true })
    })
    .catch((error) => {
      // console.error(error);
      alert(error)
      context.commit('toggle_loading', payload, { root: true })
    })
  },

  [ACTION_TYPES.UPDATE_TODO_ASYNC](context, payload) {
    context.commit('toggle_loading', payload, { root: true })
    context.commit(MUTATION_TYPES.UPDATE_SUCCESS, { success: false })
    const id = payload.todo.id
    delete payload.todo.id
    fetch(`${URL}/todos/${id}.json?user_email=${payload.user.email}`, {
      method: 'PUT',
      mode: 'cors',
      credentials: 'include',
      headers: { 
        'Content-type': 'application/json',
        'X-CSRF-Token': localStorage.csrf,
        Authorization: `Bearer ${payload.user.authentication_token}`,
      },
      body: JSON.stringify(payload.todo),
    })
    .then(res => res.json())
    .then(todo => {
      if (todo.errors) {
        context.commit(MUTATION_TYPES.UPDATE_ERRORS, { errors: todo.errors })
      } else {
        context.commit(MUTATION_TYPES.UPDATE_TODO, { todo })
        context.commit(MUTATION_TYPES.UPDATE_SUCCESS, { success: true })
      }
      context.commit('toggle_loading', payload, { root: true })
    })
    .catch((error) => {
      // console.error(error);
      alert(error)
      context.commit('toggle_loading', payload, { root: true })
    })
  },

  [ACTION_TYPES.ADD_TODO_ASYNC](context, payload) {
    context.commit('toggle_loading', payload, { root: true })
    context.commit(MUTATION_TYPES.ADD_SUCCESS, { success: false })
    fetch(`${URL}/todos.json?user_email=${payload.user.email}`, {
      method: 'POST',
      mode: 'cors',
      credentials: 'include',
      headers: { 
        'Content-type': 'application/json',
        'X-CSRF-Token': localStorage.csrf,
        Authorization: `Bearer ${payload.user.authentication_token}`,
      },
      body: JSON.stringify(payload.todo),
    })
    .then(response => {
      if (200 <= response.status < 300 && response.ok) {
        return response.json();
      } else if (response.status === 422) {
        return response.json();//validation error
      } else {
        throw new Error(`status NG ${response.status}:${response.statusText}`);
      }
    })
    .then(todo => {
      if (todo.errors) {
        context.commit(MUTATION_TYPES.ADD_ERRORS, { errors: todo.errors })
      } else {
        context.commit(MUTATION_TYPES.ADD_TODO, { todo })
        context.commit(MUTATION_TYPES.ADD_SUCCESS, { success: true })
      }
      context.commit('toggle_loading', payload, { root: true })
    })
    .catch((error) => {
      // console.error(error);
      alert(error)
      context.commit('toggle_loading', payload, { root: true })
    })
  },
};