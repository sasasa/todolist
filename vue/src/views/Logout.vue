<template>
  <form v-on:submit.prevent="submitted">
  </form>
</template>

<script>
import { LOGIN_ACTION_TYPES } from '@/store/LoginModule';
import { isAuthed } from '@/helpers/auth';

export default {
  name: 'Logout',
  data() {
    return {
    }
  },
  mounted() {
    this.submitted()
  },
  computed: {
    isAuthed() {
      return isAuthed()
    },
  },
  watch: {
    isAuthed(value) {
      if (value === false) {
        this.$router.push({ path: '/login' })
      }
    },
  },
  methods: {
    submitted() {
      this.$store.dispatch(LOGIN_ACTION_TYPES.LOGOUT_ASYNC)
    },
  },
}
</script>

<style lang="scss" scoped>
</style>