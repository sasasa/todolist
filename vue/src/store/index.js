

import Vue from 'vue';
import Vuex from 'vuex';
import createPersistedState from "vuex-persistedstate";

import * as TodoModule from './TodoModule';
import * as LoginModule from './LoginModule';

Vue.use(Vuex);

export default new Vuex.Store({
  strict: true,
  modules: {
    TodoModule,
    LoginModule,
  },
  state: {
    loading: false,
  },
  getters: {
    ['loading'](state) {
      return state.loading;
    },
  },
  mutations: {
    ['toggle_loading'](state) {
      state.loading = !state.loading 
    },
  },
  plugins: [createPersistedState()],
});

