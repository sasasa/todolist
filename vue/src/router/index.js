import Vue from 'vue'
import VueRouter from 'vue-router'
import Todos from '@/views/Todos.vue'
import Home from '@/views/Home.vue'
import Edit from '@/views/Edit.vue'
import Add from '@/views/Add.vue'
import Login from '@/views/Login.vue'
import Logout from '@/views/Logout.vue'
import Signup from '@/views/Signup.vue'
import ResetPassword from '@/views/ResetPassword.vue'
import SendEmail from '@/views/SendEmail.vue'
import { isAuthed } from '@/helpers/auth';

if (!process || process.env.NODE_ENV !== 'test') {
  Vue.use(VueRouter)
}

const routes = [
  {
    path: '/',
    name: 'home',
    component: Home,
  },
  {
    path: '/todos',
    name: 'todos',
    component: Todos,
    meta: {
      isRequestAuth: true,
    },
  },
  {
    path: '/login',
    name: 'login',
    component: Login,
  },
  {
    path: '/logout',
    name: 'logout',
    component: Logout,
  },
  {
    path: '/signup',
    name: 'signup',
    component: Signup,
  },
  {
    path: '/add',
    name: 'add',
    component: Add,
    meta: {
      isRequestAuth: true,
    },
  },
  {
    path: '/edit/:id(\\d+)',
    name: 'edit',
    component: Edit,
    props: true,
    meta: {
      isRequestAuth: true,
    },
  },
  {
    path: '/reset_password/:token',
    name: 'reset_password',
    component: ResetPassword,
    props: true,
  },
  {
    path: '/send_email',
    name: 'send_email',
    component: SendEmail,
  },
]

const router = new VueRouter({
  mode: 'history',
  base: process.env.BASE_URL,
  routes,
  scrollBehavior (to, from, savedPosition) {
    if (savedPosition) {
      return savedPosition
    } else {
      if (to.hash) {
        return { selector: to.hash }
      } else {
        return { x: 0, y: 0 }
      }
    }
  }
})
router.authed = () => {
  return isAuthed()
}
router.beforeEach((to, from, next) => {
  if(to.matched.some((route) => route.meta.isRequestAuth) && !router.authed()) {
    next({ path: '/login', query: { backuri: to.fullPath } })
  } else {
    next()
  }
})

export default router
