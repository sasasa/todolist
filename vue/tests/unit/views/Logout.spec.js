import { shallowMount, mount, createLocalVue } from '@vue/test-utils'
import flushPromises from 'flush-promises'
import VueRouter from 'vue-router'
import router from '../router'
import Vuex from 'vuex'
import Logout from '@/views/Logout'
import * as TodoModule from '../store/TodoModuleMock'
import * as LoginModule from '../store/LoginModuleMock'




describe('Logout view', () => {
  let localVue
  let store
  let $router

  beforeEach(() => {
    localVue = createLocalVue()
    // localVue.use(VueRouter)
    localVue.use(Vuex)
    store = new Vuex.Store({
      modules: {
        TodoModule,
        LoginModule,
      }
    })
    $router = {
      push: jest.fn(),
    }
    // LoginModule.actions[LoginModule.ACTION_TYPES.LOGOUT_ASYNC] = jest.fn()
  })
  
  describe('methods', () => {
    it('submitted', () => {
      localVue.use(VueRouter)
      const wrapper = shallowMount(Logout, {
        store,
        router,
        localVue,
      })
      // 明示的に呼ばずとも mounted から呼ばれる
      // 順番を帰るとテストが失敗するためテストケースを動かさない
      expect(LoginModule.actions[LoginModule.ACTION_TYPES.LOGOUT_ASYNC].mock.calls).toHaveLength(1)
    })
  })

  it('mount', () => {
    localVue.use(VueRouter)
    const wrapper = shallowMount(Logout, {
      store,
      router,
      localVue,
    })
    expect(wrapper.isVueInstance()).toBe(true)
  })
  describe('computed -> watch', () => {
    it('isAuthed false', async() => {
      Logout.computed.isAuthed = () => LoginModule.getters[LoginModule.GETTER_TYPES.LOGIN](LoginModule.state)
      const wrapper = shallowMount(Logout, {
        store,
        // router,
        localVue,
        mocks: {
          $router,
        },
        stubs: ['router-link', 'router-view']
      })
      LoginModule.mutations[LoginModule.MUTATION_TYPES.LOGIN](LoginModule.state, { login: true })
      expect(wrapper.vm.isAuthed).toBe(true)
      LoginModule.mutations[LoginModule.MUTATION_TYPES.LOGIN](LoginModule.state, { login: false })
      expect(wrapper.vm.isAuthed).toBe(false)
      
      // ここでwatchの処理が動いたかどうかをテストする
      await flushPromises()

      expect($router.push.mock.calls).toHaveLength(1)
      expect($router.push.mock.calls[0][0]).toEqual({ path: '/login' })
      
    })

  })

  describe('template', () => {
    it('@submit.prevent="submitted"', async() => {
      localVue.use(VueRouter)
      const mock = jest.fn()
      const wrapper = shallowMount(Logout, {
        store,
        router,
        localVue,
      })
      wrapper.setMethods({
        submitted: mock
      })

      wrapper.find('form').trigger('submit.prevent')
      expect(mock).toHaveBeenCalled()
    })

    it('snapshot', () => {
      localVue.use(VueRouter)
      const wrapper = shallowMount(Logout, {
        store,
        router,
        localVue,
      })
      expect(wrapper.vm.$el).toMatchSnapshot()
    })
  })
})