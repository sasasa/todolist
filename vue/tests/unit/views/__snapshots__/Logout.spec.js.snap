// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logout view template snapshot 1`] = `<form />`;
